### ISSUE TITLE
I make my own Network to train machine and data is used Mnist But it's give Low accuracy and somtimes my softmax code not working as well.. so please tell me what's wrong with my code.... 

### ISSUE THREAD

**Issue by Sunil1997** [tensorflow/tensorflow#21828]
**somtimes softmax give null value after updating weights so it's throw divison by zero error please check my code and tell me what's wrong with it.
Used: python 3.6.5**

import numpy as np
import gzip
import math
IMAGE_SIZE = 28

def sigmoid(x):
    # return (2 / (1 + np.exp(-2 * (x)))) - 1
    return 1.0 / (1.0 + np.exp(-x))

def derivation_sigmoid(x):
    # return 1 - (x ** 2)
    return x * (1.0 - x)

def softmax(x):
    # print np.sum(np.exp(x))
    return (np.exp(x)) / (np.sum(np.exp(x)))

def extract_data(filename, num_images):
  with gzip.open(filename) as bytestream:
    bytestream.read(16)
    buf = bytestream.read(IMAGE_SIZE * IMAGE_SIZE * num_images)
    data = np.frombuffer(buf, dtype=np.uint8).astype(np.float32)
    data = data.reshape(num_images, IMAGE_SIZE, IMAGE_SIZE, 1)
    return data

def extract_labels(filename, num_images):
  with gzip.open(filename) as bytestream:
    bytestream.read(8)
    buf = bytestream.read(1 * num_images)
    labels = np.frombuffer(buf, dtype=np.uint8).astype(np.int64)
  return labels

def derivation_cross(x, y):
    return -1 * ((x * (1 / y)) + (1 - x) * (1 / (1 - y)))

def derivation_softmax(x):
    tem = sum(np.exp(x)) ** 2
    # print tem
    tem1 = sum(np.exp(x))
    a = []
    for i_ in range(len(x)):
        t1 = (np.exp(x[i_]) * (tem1 - np.exp(x[i_]))) / tem
        a.append(t1)
    return a

train_data_filename = "/home/rootpranav/Downloads/train-images-idx3-ubyte.gz"
train_data = extract_data(train_data_filename, 60000)

train_data_label = "/home/rootpranav/Downloads/train-labels-idx1-ubyte.gz"
tarin_data_label_1 = extract_labels(train_data_label, 6000)

input_array = []
counter = 0

input_neurons = train_data.shape[1] * train_data.shape[1]   # Number of Feature
out_neurons = 10
hidden_layer_nurons = 15
lr = 0.1

bias_hidden = np.ones((15, 1))
bias_output = np.ones((10, 1))

wih = np.ones((15,784))
who = np.ones((10, 15))

def neural_network(wih, who, train_data, tarin_data_label_1):
    for e_ in range(0, 1000):
        for da_ in range(900, 980):
            
            c_error = []
            m1 = np.matrix((train_data[da_]))
            m2 = np.reshape(m1, (784, 1))
            m2 = ((m2 * 1) / 255)
            
            hidden_sigmoid = sigmoid((np.dot(wih, m2)))
            output = (np.dot(who, hidden_sigmoid))
            output_softmax = softmax((np.dot(who, hidden_sigmoid)))

            index = tarin_data_label_1[da_]
            for i in range(0, 10):
                if i != index:
                    c_error.append([0])
                else:
                    c_error.append([1])
            
            a1 = np.array(c_error)
            a2 = np.array(output_softmax)

            # This is for Hidden -> Output change weight
            cross_entropy_derivation = derivation_cross(a1, a2)

            # print cross_entropy_derivation
            softmax_derivation = derivation_softmax(output)
            sigmoid_derivation = derivation_softmax(hidden_sigmoid)
            sigmoid_derivation = np.reshape(sigmoid_derivation, (15, 1))

            m = np.reshape(softmax_derivation, (10, 1))
            d_output = cross_entropy_derivation * m

            who -= (d_output.dot(hidden_sigmoid.T))

            # this is for Hidden -> Input

            d_hidden = who.T.dot(d_output)
            d1 =  d_hidden * sigmoid_derivation
            wih -= d1.dot(m2.T)
            # print who[0]
            test = np.argmax(output_softmax)
            print test, index

neural_network(wih, who, train_data, tarin_data_label_1)

**Comment by nkcr7** [comment]
First I guess this should not be posted here. So you initialized the weights to ones like `wih = np.ones((15,784))`, which makes the network symmetric and hard to train. You should initialize the weights to small small random values.

**Comment by Sunil1997** [comment]
Already Assigned same things check my code @nkcr7

**Comment by Sunil1997** [comment]
when i start my above code after **1-2 iteration**,my **softmax function** given **Nan** value to me so **derivation of softmax** give me **division by Zero Error**.How that i don't know. If @nkcr7 @karmel know then tell me what can i do??

**Comment by Keshav-Aggarwal** [comment]
You are posting it at wrong platform. It would be better if you can post it on stackoverflow as this is related to tensorflow related issues.

**Comment by karmel** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!
